### PR TITLE
as.integer()

### LINKED ISSUE #82: Add `as.integer()`
Currently:

``` r
data.frame(a = "X") |>
  duckplyr::as_duckplyr_df() |>
  dplyr::mutate(b = as.integer(a))
#> materializing:
#> ---------------------
#> --- Relation Tree ---
#> ---------------------
#> Projection [a as a, as.integer(a) as b]
#>   r_dataframe_scan(0x11297e1d0)
#> 
#> ---------------------
#> -- Result Columns  --
#> ---------------------
#> - a (VARCHAR)
#> - b (INTEGER)
#> Error: Error evaluating duckdb query: Conversion Error: Could not convert string 'X' to INT32
```

<sup>Created on 2024-04-26 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Translated to:

``` r
duckdb <- asNamespace("duckdb")
con <- DBI::dbConnect(duckdb::duckdb())
experimental <- FALSE
invisible(DBI::dbExecute(con, 'CREATE MACRO "as.integer"(x) AS CAST(x AS int32)'))
df1 <- data.frame(a = "X")

rel1 <- duckdb$rel_from_df(con, df1, experimental = experimental)
rel2 <- duckdb$rel_project(
  rel1,
  list(
    {
      tmp_expr <- duckdb$expr_reference("a")
      duckdb$expr_set_alias(tmp_expr, "a")
      tmp_expr
    },
    {
      tmp_expr <- duckdb$expr_function("as.integer", list(duckdb$expr_reference("a")))
      duckdb$expr_set_alias(tmp_expr, "b")
      tmp_expr
    }
  )
)
rel2
#> DuckDB Relation: 
#> ---------------------
#> --- Relation Tree ---
#> ---------------------
#> Projection [a as a, as.integer(a) as b]
#>   r_dataframe_scan(0x10bee1a30)
#> 
#> ---------------------
#> -- Result Columns  --
#> ---------------------
#> - a (VARCHAR)
#> - b (INTEGER)
duckdb$rel_to_altrep(rel2)
#> Error: Error evaluating duckdb query: Conversion Error: Could not convert string 'X' to INT32
```

<sup>Created on 2024-04-26 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Expected behavior: same as R. We need to think about how to propagate the warning.

``` r
data.frame(a = "X") |>
  dplyr::mutate(b = as.integer(a))
#> Warning: There was 1 warning in `dplyr::mutate()`.
#> ℹ In argument: `b = as.integer(a)`.
#> Caused by warning:
#> ! NAs introduced by coercion
#>   a  b
#> 1 X NA
```

<sup>Created on 2024-04-26 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #82

### CODE REVIEW

**Review by krlmlr — APPROVED** [review]
Thanks, looks good. Have we moved away from code generation?

File: src/as_integer.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	if (isnan(input) || input > std::numeric_limits<int32_t>::max() || input < std::numeric_limits<int32_t>::min() ) {
+		mask.SetInvalid(idx);
+	}
+
+	return static_cast<int32_t>(input);
+}
+
+}
+
+#define AS_INTEGER(__TYPE__) ScalarFunction(                                                      \
```
Comment (krlmlr) [inline src/as_integer.cpp:30]: Does this have to be a macro? Could it be a template function?

Comment (romainfrancois) [reply src/as_integer.cpp]: ✅

**Comment by romainfrancois** [comment]
> Have we moved away from code generation?

Not necessarily. It feels a little easier to lean on `rfuns_project()` to test, but I could code-generate them if you prefer.

**Comment by romainfrancois** [comment]
At the moment, I don't know how to handle the generation of the warnings. I'll digg in the `ScalarFunction` class, or perhaps I can do something with the `ExpressionState &state` or the `Vector &result`: 

``` r
library(duckdbrfuns)

rfuns_as.integer("x")
#> [1] NA
as.integer("x")
#> Warning: NAs introduced by coercion
#> [1] NA
```

<sup>Created on 2024-05-01 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Pull request merged**